google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -321,50 +348,385 @@ private boolean isAlive() {
+        // If enableExactlyOnceDelivery is true:
+        // 1) make a response future
+        // 2) add it to our map
+        // 3) add it to the logging callback
+        // 4) populate the ackIds for the request
+        SettableApiFuture<Map<String, String>> responseFuture = SettableApiFuture.create();
+        for (MessageDispatcher.AckWithMessageFuture ackWithMessageFuture :
+            ackIdWithMessageFutureInRequest) {
+          futureMap.put(ackWithMessageFuture.ackId, responseFuture);
+          ackIdsInRequest.add(ackWithMessageFuture.ackId);
+        }
+        loggingCallback = getLoggingCallback(responseFuture);
+      } else {
+        // else, we just need to populate the ackIds for the request
+        for (MessageDispatcher.AckWithMessageFuture ackWithMessageFuture :
+                ackIdWithMessageFutureInRequest) {
+          ackIdsInRequest.add(ackWithMessageFuture.ackId);
+        }
+        loggingCallback = getLoggingCallback();
+      }
       ApiFuture<Empty> future =
-          stub.acknowledgeCallable()
+          subscriberStub
+              .acknowledgeCallable()
               .futureCall(
                   AcknowledgeRequest.newBuilder()
                       .setSubscription(subscription)
-                      .addAllAckIds(idChunk)
+                      .addAllAckIds(ackIdsInRequest)
                       .build());
       ApiFutures.addCallback(future, loggingCallback, directExecutor());
       pendingOperations++;
     }
-
     ackOperationsWaiter.incrementPendingCount(pendingOperations);
+    return futureMap;
+  }
+
+  private List<MessageDispatcher.AckWithMessageFuture> processAckFutures(
+      List<MessageDispatcher.AckWithMessageFuture> ackIdToSendWithMessageFuture,
+      Map<String, SettableApiFuture<Map<String, String>>> futureMap) {
+    List<MessageDispatcher.AckWithMessageFuture> retryAckIdsWithMessageFuture = new ArrayList<>();
+    ackIdToSendWithMessageFuture.forEach(
+        (ackIdWithMessageFuture) -> {
+          String ackId = ackIdWithMessageFuture.ackId;
+          SettableApiFuture<Map<String, String>> future = futureMap.get(ackId);
+          try {
+            // Blocking operation to check the response of the ack request
+            Map<String, String> metadataMap = future.get();
+            if (metadataMap.containsKey(ackId)) {
+              String errorMessage = metadataMap.get(ackId);
+              if (errorMessage.startsWith(TRANSIENT_ERROR_METADATA_PREFIX)) {
+                // Retry and do not set message future
+                logger.log(Level.WARNING, "Transient error message, will resend", errorMessage);
+                retryAckIdsWithMessageFuture.add(ackIdWithMessageFuture);
+              } else if (errorMessage.startsWith(PERMANENT_ERROR_METADATA_PREFIX)) {
+                logger.log(Level.WARNING, "Permanent error message, will not resend", errorMessage);
+                ackIdWithMessageFuture.messageFuture.set(AckResponse.INVALID);
```

Comment:
Not every permanent error should be counted as invalid (it might be the case but we can expand the list of permanent errors later). Please check the full string for the permanent invalid ack id. And if it's a permanent error but not invalid, assign it to OTHER.